File: java/P12.java
// P12 Decode a run-length encoded list.

import core.List;
import core.Pair;

import static core.List.*;
import static core.Pair.*;

class P12 {
    // Imperative, list of objects
    @SuppressWarnings("unchecked")
    List f1(List list) {
        List acc = nil();
        for (Object each : list) {
            if (each instanceof Pair) {
                Pair<Integer, Object> pair = (Pair<Integer, Object>) each;
                acc = concat(List.replicate(pair.getA(), pair.getB()), acc);
            } else {
                acc = cons(each, acc);
            }
        }
        return acc.reverse();
    }
}

File: clojure/P12.clj
; P12 Decode a run-length encoded list.

; Mapping
(defn f1 [l]
  (defn f [x] (if (seq? x) (repeat (first x) (second x)) x))
  (flatten (map f l)))

; Recursion
(defn f2 [l]
  (if (seq l)
    (let [h (first l) t (rest l)]
      (if (seq? h)
        (concat (repeat (first h) (second h)) (f2 t))
        (cons h (f2 t))))
    '()))